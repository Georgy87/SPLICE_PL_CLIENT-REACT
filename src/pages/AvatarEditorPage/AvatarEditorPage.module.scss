.root {
	display: flex;
	justify-content: center;
}

.downloadAvatar {
    width: 500px;
	height: 300px;
    margin: 0 auto;
	margin-top: 80px;
    text-align: center;
	border-style: none;
	border-radius: 4px;
	cursor: pointer;
	box-shadow: inset 2px 3px 10px rgb(228, 241, 244);
}
