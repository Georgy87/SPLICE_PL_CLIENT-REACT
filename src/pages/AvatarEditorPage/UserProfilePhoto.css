.root {
	display: flex;
	justify-content: space-between;
	align-items: center;
	width: 900px;
}

.downloadAvatar {
	display: flex;
	justify-content: center;
	align-items: center;
	width: 500px;
	height: 300px;
	margin: 0 auto;
	margin-top: 80px;
	text-align: center;
	border-style: none;
	border-radius: 4px;
	cursor: pointer;
	box-shadow: inset 2px 3px 10px rgb(228, 241, 244);
}

.download-image {
	width: 500px;
	height: 300px;
	margin: 0 auto;
	margin-top: 80px;
	border-style: none;
	border-radius: 4px;
	text-align: center;
	cursor: pointer;
	box-shadow: inset 2px 3px 10px rgb(228, 241, 244);
}

.download-file-image {
	margin: 0 auto;
	margin-top: 200px;
}

.download-image p {
	font-weight: 600;
}

.preview-avatar {
	display: flex;
	align-items: center;
	flex-direction: column;
}

.preview-avatar p {
	margin-bottom: 10px;
	font-weight: 600;
}

.dropzone {
	display: flex;
	justify-content: center;
	align-items: center;
	width: 200px;
	height: 200px;
}

.ReactCrop {
	position: relative;
	display: flex;
	cursor: crosshair;
	overflow: hidden;
	max-width: 100%;
	background-color: #000;
	margin-top: 80px;
}

.ReactCrop:focus {
	outline: none;
}

.ReactCrop--disabled {
	cursor: inherit;
}

.ReactCrop__image {
	display: block;
	max-width: 100%;
	max-height: -webkit-stretch;
	max-height: -moz-available;
	max-height: stretch;
}

.ReactCrop--crop-invisible .ReactCrop__image {
	opacity: 0.5;
}

.ReactCrop__crop-selection {
	position: absolute;
	top: 0;
	left: 0;
	transform: translate3d(0, 0, 0);
	box-sizing: border-box;
	cursor: move;
	box-shadow: 0 0 0 9999em rgba(0, 0, 0, 0.5);
	border: 1px solid;
	border-image-repeat: repeat;
}

.ReactCrop--disabled .ReactCrop__crop-selection {
	cursor: inherit;
}

.ReactCrop__drag-handle {
	position: absolute;
	width: 9px;
	height: 9px;
	background-color: rgba(0, 0, 0, 0.2);
	border: 1px solid rgba(255, 255, 255, 0.7);
	box-sizing: border-box;
	outline: 1px solid transparent;
}

.ReactCrop .ord-nw {
	top: 0;
	left: 0;
	margin-top: -5px;
	margin-left: -5px;
	cursor: nw-resize;
}

.ReactCrop .ord-n {
	top: 0;
	left: 50%;
	margin-top: -5px;
	margin-left: -5px;
	cursor: n-resize;
}

.ReactCrop .ord-ne {
	top: 0;
	right: 0;
	margin-top: -5px;
	margin-right: -5px;
	cursor: ne-resize;
}

.ReactCrop .ord-e {
	top: 50%;
	right: 0;
	margin-top: -5px;
	margin-right: -5px;
	cursor: e-resize;
}

.ReactCrop .ord-se {
	bottom: 0;
	right: 0;
	margin-bottom: -5px;
	margin-right: -5px;
	cursor: se-resize;
}
.ReactCrop .ord-s {
	bottom: 0;
	left: 50%;
	margin-bottom: -5px;
	margin-left: -5px;
	cursor: s-resize;
}
.ReactCrop .ord-sw {
	bottom: 0;
	left: 0;
	margin-bottom: -5px;
	margin-left: -5px;
	cursor: sw-resize;
}
.ReactCrop .ord-w {
	top: 50%;
	left: 0;
	margin-top: -5px;
	margin-left: -5px;
	cursor: w-resize;
}
.ReactCrop__disabled .ReactCrop__drag-handle {
	cursor: inherit;
}
.ReactCrop__drag-bar {
	position: absolute;
}
.ReactCrop__drag-bar.ord-n {
	top: 0;
	left: 0;
	width: 100%;
	height: 6px;
	margin-top: -3px;
}
.ReactCrop__drag-bar.ord-e {
	right: 0;
	top: 0;
	width: 6px;
	height: 100%;
	margin-right: -3px;
}
.ReactCrop__drag-bar.ord-s {
	bottom: 0;
	left: 0;
	width: 100%;
	height: 6px;
	margin-bottom: -3px;
}
.ReactCrop__drag-bar.ord-w {
	top: 0;
	left: 0;
	width: 6px;
	height: 100%;
	margin-left: -3px;
}
.ReactCrop--new-crop .ReactCrop__drag-bar,
.ReactCrop--new-crop .ReactCrop__drag-handle,
.ReactCrop--fixed-aspect .ReactCrop__drag-bar {
	display: none;
}
.ReactCrop--fixed-aspect .ReactCrop__drag-handle.ord-n,
.ReactCrop--fixed-aspect .ReactCrop__drag-handle.ord-e,
.ReactCrop--fixed-aspect .ReactCrop__drag-handle.ord-s,
.ReactCrop--fixed-aspect .ReactCrop__drag-handle.ord-w {
	display: none;
}
@media (max-width: 768px), (pointer: coarse) {
	.root {
		flex-direction: column;
	}
	.ReactCrop__drag-handle {
		width: 17px;
		height: 17px;
	}
	.ReactCrop .ord-nw {
		margin-top: -9px;
		margin-left: -9px;
	}
	.ReactCrop .ord-n {
		margin-top: -9px;
		margin-left: -9px;
	}
	.ReactCrop .ord-ne {
		margin-top: -9px;
		margin-right: -9px;
	}
	.ReactCrop .ord-e {
		margin-top: -9px;
		margin-right: -9px;
	}
	.ReactCrop .ord-se {
		margin-bottom: -9px;
		margin-right: -9px;
	}
	.ReactCrop .ord-s {
		margin-bottom: -9px;
		margin-left: -9px;
	}
	.ReactCrop .ord-sw {
		margin-bottom: -9px;
		margin-left: -9px;
	}
	.ReactCrop .ord-w {
		margin-top: -9px;
		margin-left: -9px;
	}
	.ReactCrop__drag-bar.ord-n {
		height: 14px;
		margin-top: -7px;
	}
	.ReactCrop__drag-bar.ord-e {
		width: 14px;
		margin-right: -7px;
	}
	.ReactCrop__drag-bar.ord-s {
		height: 14px;
		margin-bottom: -7px;
	}
	.ReactCrop__drag-bar.ord-w {
		width: 14px;
		margin-left: -7px;
	}
}
.drag-container {
	display: flex;
	width: 85%;
	flex-direction: column;
}
.photo-titles {
	display: flex;
	justify-content: center;
	flex-direction: column;
	align-items: center;
	width: 100%;
	margin: 0 auto;
	border-bottom: 1px solid rgb(211, 209, 209);
	border-left: none;
}
.photo-titles p {
	font-size: 20px;
	margin: 10px;
	margin-top: 20px;
	font-family: var(--myFontFamily);
}
.photo-descr {
	font-family: var(--myFontFamily);
	margin: 10px;
}
.photo-add-container {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
}
#file-input {
	display: none;
}

.file-upload-wrapper {
	display: flex;
	margin-top: 15px;
	height: 30px;
}
.file-upload-wrapper label {
	align-items: center;
	display: flex;
}
.dropzone-text {
	margin-top: 25%;
	color: rgb(207, 204, 204);
	font-family: var(--myFontFamily);
}
.ReactCrop--fixed-aspect .ReactCrop__image {
	width: 500px;
	/* min-height: 400px; */
	margin: 0 auto;
}
.photo-avatar-btns {
	display: flex;
	margin-top: 10px;
	justify-content: center;
}
.photo-avatar-btns button {
	display: flex;
	background-color: white;
	margin-top: 10px;
	border-radius: 3px;
	height: 40px;
	align-items: center;
	padding: 15px;
}
.photo-avatar-clear-btn {
	margin-left: 5px;
	border: 1px solid rgb(201, 76, 67);
	color: rgb(201, 76, 67);
}
.photo-avatar-save-btn {
	margin-right: 5px;
	border: 1px solid rgb(86, 192, 206);
	color: rgb(86, 192, 206);
}
.choose-file-btns {
	display: flex;
	margin-bottom: 20px;
	margin-top: 50px;
}
.choose-file-btns #choose-file,
.no-file-selected {
	display: flex;
	margin-top: 10px;
	align-items: center;
	padding: 15px;
	height: 10px;
}
#choose-file {
	margin-left: 5px;
	background-color: white;
	border: 1px solid rgb(86, 209, 225);
	color: rgb(86, 209, 225);
	cursor: pointer;
	border-radius: 5px 0 0 5px;
	font-family: var(--myFontFamily);
}
.no-file-selected {
	border: 1px solid rgb(192, 187, 187);
	border-left: none;
	color: rgb(117, 117, 115);
	background-color: rgb(228, 241, 238);
	width: 300px;
	border-radius: 0 5px 5px 0;
	font-family: var(--myFontFamily);
}

@media (max-width: 900px), (pointer: coarse) {
	.download-image {
		display: none;
	}
}

@media (max-width: 414px), (pointer: coarse) {
	.download-image {
		display: none;
	}
}
