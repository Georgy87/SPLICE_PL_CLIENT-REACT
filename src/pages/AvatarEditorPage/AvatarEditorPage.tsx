import React, { useCallback, useEffect, useRef, useState } from 'react';
import ReactCrop, { Crop } from 'react-image-crop';

import { useDropzone } from '../../hooks/useDropzone';
import { avatarService } from '../../utils/avatarService';

import './UserProfilePhoto.css';

export const AvatarEditorPage = () => {
	const imgRef = useRef<HTMLImageElement | HTMLCanvasElement | null>(null);
	const previewCanvasRef = useRef<HTMLCanvasElement | null>(null);

	const [avatarState, setAvatarState] = useState<{
		imgSrc: any;
		imgSrcExt: string | ArrayBuffer | null | undefined;
	}>({
		imgSrc: null,
		imgSrcExt: null,
	});

	const [crop, setCrop] = useState<any>({ unit: '%', width: 30, aspect: 1 / 1 });
	const [completedCrop, setCompletedCrop] = useState(null);

	const { drag, setDrag, dragEnter, dragLeave } = useDropzone();

	const handleOnDrop = (e: React.DragEvent<HTMLDivElement>) => {
		e.preventDefault();
		e.stopPropagation();
		const eventData = (e as React.DragEvent).dataTransfer;
		const files = [eventData.files];

		if (files && files.length > 0) {
			const currentFile = files[0];

			const reader = new FileReader();
			reader.addEventListener(
				'load',
				() => {
					const myResult = reader.result;

					setAvatarState({
						imgSrc: myResult,
						imgSrcExt: avatarService.extractImageFileExtensionFromBase64(myResult),
					});
				},
				false,
			);
			reader.readAsDataURL(currentFile[0]);
		}
	};

	const onLoad = useCallback((img) => {
		imgRef.current = img;
	}, []);

	useEffect(() => {
		if (!completedCrop || !previewCanvasRef.current || !imgRef) {
			return;
		}

		const image: any = imgRef.current;

		if (!image) return;
		
		const canvas = previewCanvasRef.current;
		const crop: any = completedCrop;

		const scaleX = image.naturalWidth / image.width;
		const scaleY = image.naturalHeight / image.height;
		
		const ctx: CanvasRenderingContext2D | null  = canvas.getContext('2d');
		const pixelRatio = window.devicePixelRatio;

		canvas.width = crop.width * pixelRatio * scaleX;
		canvas.height = crop.height * pixelRatio * scaleY;

		if (!ctx) return;

		ctx.setTransform(pixelRatio, 0, 0, pixelRatio, 0, 0);
		ctx.imageSmoothingQuality = 'high';

		ctx.drawImage(
			image,
			crop.x * scaleX,
			crop.y * scaleY,
			crop.width * scaleX,
			crop.height * scaleY,
			0,
			0,
			crop.width * scaleX,
			crop.height * scaleY,
		);
	}, [completedCrop]);

	return (
		<div>
			<div className='photo-add-container'>
				{avatarState.imgSrc ? (
					<>
						<ReactCrop
							src={avatarState.imgSrc}
							crop={crop}
							onImageLoaded={onLoad}
							onChange={(c) => setCrop(c)}
							onComplete={(c: any) => setCompletedCrop(c)}
						/>
						<br />
						<p
							style={{
								textAlign: 'center',
								fontFamily: 'var(--myFontFamily)',
							}}
						>
							Предпросмотр аватара
						</p>
						<canvas
							ref={previewCanvasRef}
							style={{
								width: '200px',
								height: '200px',
								display: 'block',
								margin: '0 auto',
								borderRadius: '100%',
							}}
						></canvas>
					</>
				) : (
					<div
						style={{
							borderStyle: 'none',
							width: '500px',
							height: '300px',
							borderRadius: '4px',
							textAlign: 'center',
							margin: '0 auto',
							marginTop: '80px',
							cursor: 'pointer',
							boxShadow: 'inset 2px 3px 10px rgb(228, 241, 244)',
							// backgroundColor: "rgb(243, 245, 244)"
						}}
						onDragEnter={dragEnter}
						onDragLeave={dragLeave}
						onDragOver={dragEnter}
						onDrop={handleOnDrop}
					>
						<p>Download Image</p>
					</div>
				)}
			</div>
		</div>
	);
};
