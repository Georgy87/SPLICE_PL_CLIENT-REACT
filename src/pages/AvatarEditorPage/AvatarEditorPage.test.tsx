import { render } from '@testing-library/react';
import { renderTest } from '../../tests/helpers/RenderTestApp';

import { AvatarEditorPage } from './AvatarEditorPage';

describe('AvatarEditorPage component', () => {
	test('AvatarEditorPage snapshot', async () => {
		const avatarEditorPage = render(renderTest(<AvatarEditorPage />));
		expect(avatarEditorPage).toMatchSnapshot();
	});
});
