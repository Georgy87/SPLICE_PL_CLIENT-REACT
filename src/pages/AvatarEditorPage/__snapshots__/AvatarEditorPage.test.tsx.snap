// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarEditorPage component AvatarEditorPage snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
        data-testid="avatar-editor-page"
      >
        <div
          class="download-image"
        >
          <p>
            Download Image
          </p>
        </div>
        <div
          class="download-file-image"
        >
          <input
            style="display: none;"
            type="file"
          />
          <div
            class="platformBtn"
          >
            <button
              class="platformBtn download"
            >
              <div
                class="iconContainer"
              >
                <span
                  aria-label="upload"
                  class="anticon anticon-upload"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="upload"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                    />
                  </svg>
                </span>
              </div>
              UPLOAD AVATAR
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
      data-testid="avatar-editor-page"
    >
      <div
        class="download-image"
      >
        <p>
          Download Image
        </p>
      </div>
      <div
        class="download-file-image"
      >
        <input
          style="display: none;"
          type="file"
        />
        <div
          class="platformBtn"
        >
          <button
            class="platformBtn download"
          >
            <div
              class="iconContainer"
            >
              <span
                aria-label="upload"
                class="anticon anticon-upload"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="upload"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </div>
            UPLOAD AVATAR
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
